pilot/pkg/model/push_context.go

```diff
@@ -1440,27 +1440,21 @@ func (ps *PushContext) initSidecarScopes(env *Environment) error {
 
 	sidecarNum := len(sidecarConfigs)
 	sidecarConfigs = make([]config.Config, 0, sidecarNum)
+	// sidecars with selector take preference
 	sidecarConfigs = append(sidecarConfigs, sidecarConfigWithSelector...)
 	sidecarConfigs = append(sidecarConfigs, sidecarConfigWithoutSelector...)
 
-	ps.sidecarsByNamespace = make(map[string][]*SidecarScope, sidecarNum)
-	for _, sidecarConfig := range sidecarConfigs {
-		sidecarConfig := sidecarConfig
-		ps.sidecarsByNamespace[sidecarConfig.Namespace] = append(ps.sidecarsByNamespace[sidecarConfig.Namespace],
-			ConvertToSidecarScope(ps, &sidecarConfig, sidecarConfig.Namespace))
-	}
-
 	// Hold reference root namespace's sidecar config
 	// Root namespace can have only one sidecar config object
 	// Currently we expect that it has no workloadSelectors
 	var rootNSConfig *config.Config
-	if ps.Mesh.RootNamespace != "" {
-		for i, sidecarConfig := range sidecarConfigs {
-			if sidecarConfig.Namespace == ps.Mesh.RootNamespace &&
-				sidecarConfig.Spec.(*networking.Sidecar).WorkloadSelector == nil {
-				rootNSConfig = &sidecarConfigs[i]
-				break
-			}
+	ps.sidecarsByNamespace = make(map[string][]*SidecarScope, sidecarNum)
+	for i, sidecarConfig := range sidecarConfigs {
+		ps.sidecarsByNamespace[sidecarConfig.Namespace] = append(ps.sidecarsByNamespace[sidecarConfig.Namespace],
```

Comment:
sidecarConfig := sidecarConfig was missing. Not sure why it was added in the first place. But bringing this out

Reply:
Yes I think its a bug since we take the pointer of sidecarConfig (https://medium.com/swlh/use-pointer-of-for-range-loop-variable-in-go-3d3481f7ffc9). I hate this "feature" of go...

Reply:
I deliberately removed it, in the loop we donot store the sidecarConfig address, just read its fields, especially the spec, which is a pointer of `sidecar`

Reply:
are you sure in ConvertToSidecarScope it's not stored?

Reply:
yes, it stores the `sidecar` field

Reply:
sidecar := sidecarConfig.Spec.(*networking.Sidecar)

Reply:
this looks fine to me